src/gui/osutils/nixutils/NixUtils.cpp

```diff
@@ -134,3 +151,119 @@ bool NixUtils::isCapslockEnabled()
 
     return false;
 }
+
+void NixUtils::registerNativeEventFilter()
+{
+    qApp->installNativeEventFilter(this);
+}
+
+bool NixUtils::nativeEventFilter(const QByteArray& eventType, void* message, long*)
+{
+    if (eventType != QByteArrayLiteral("xcb_generic_event_t")) {
+        return false;
+    }
+
+    auto* genericEvent = static_cast<xcb_generic_event_t*>(message);
+    quint8 type = genericEvent->response_type & 0x7f;
+
+    if (type == XCB_KEY_PRESS) {
+        auto* keyPressEvent = static_cast<xcb_key_press_event_t*>(message);
+        auto modifierMask = ControlMask | ShiftMask | Mod1Mask | Mod4Mask;
+        return triggerGlobalShortcut(keyPressEvent->detail, keyPressEvent->state & modifierMask);
+    } else if (type == XCB_MAPPING_NOTIFY) {
+        auto* mappingNotifyEvent = static_cast<xcb_mapping_notify_event_t*>(message);
+        if (mappingNotifyEvent->request == XCB_MAPPING_KEYBOARD
+            || mappingNotifyEvent->request == XCB_MAPPING_MODIFIER) {
+            XMappingEvent xMappingEvent;
+            memset(&xMappingEvent, 0, sizeof(xMappingEvent));
+            xMappingEvent.type = MappingNotify;
+            xMappingEvent.display = dpy;
+            if (mappingNotifyEvent->request == XCB_MAPPING_KEYBOARD) {
+                xMappingEvent.request = MappingKeyboard;
+            } else {
+                xMappingEvent.request = MappingModifier;
+            }
+            xMappingEvent.first_keycode = mappingNotifyEvent->first_keycode;
+            xMappingEvent.count = mappingNotifyEvent->count;
+            XRefreshKeyboardMapping(&xMappingEvent);
+            // Notify listeners that the keymap has changed
+            emit keymapChanged();
+        }
+    }
+
+    return false;
+}
+
+bool NixUtils::triggerGlobalShortcut(uint keycode, uint modifiers)
```

Comment:
If this is directly a part of NixUtils, how are we handling Wayland in the future? That would be an entirely new implementation. It could make sense to add the necessary abstractions already. I believe it should also be possible to build without X11 headers present on the system.

Reply:
Perhaps we need an X11Utils and WaylandUtils

Reply:
I'd rather think we need to add another abstraction layer there. As I see it, NixUtils will grow with additional features that are specific to certain DEs and we don't want to duplicate all that unnecessarily.

Reply:
I also think that the possibility to handle features specific to DEs is necessary. And FYI what a DE-specific implementation looks like: I tried to [implement](https://github.com/Aetf/keepassxc/tree/feature/autotype-wayland) auto-type on KDE Wayland a while ago.